### ISSUE TITLE
Improve logging for IPAMD when it is unable to reach the ec2 endpoint to get the ENI info by executing the DescribeNetworkInterfaces API call.

### ISSUE THREAD

**Issue by skkumsi** [aws/amazon-vpc-cni-k8s#1558]
**What would you like to be added**:

Improve logging for IPAMD when it is unable to communicate with AWS EC2 Service endpoint. According to the current behavior, the plugin just stops logging further or it continues to start up the initialization process until it successfully starts up. This behavior is observed with both 1.7.10 and 1.8.0 versions of AWS VPC CNI Plugin.

**Why is this needed**:

During the initialization process, AWS VPC CNI plugin tries to communicate with the EC2 service endpoint and when it is unable to reach the endpoint, the logging it provides is not very helpful in identifying the underlying issue for faster issue resolution.

**Sample use case**: 
* The issue surfaces when the VPC CNI plugin (aws-node pod) as part of its start up process, tries to communicate with EC2 service endpoint to execute the DescribeNetworkInterfaces API call to fetch the underlying worker node ENI info and fails. 
> Note: The worker node VPC has also got a VPC endpoint created for EC2 service.
* The EC2 API call fails because the VPC in which the worker nodes are launched is configured with Custom DNS servers without the required conditional forwarding, thereby making the EC2 service endpoint not reachable.

**Logging observed:** 
The current logging available does not really help with figuring out the issue with not being able to execute the DescribeNetworkInterfaces API call.
```
{"level":"info","ts":"2021-06-14T10:57:32.084Z","caller":"aws-k8s-agent/main.go:28","msg":"Starting L-IPAMD v1.7.10  ..."}
{"level":"info","ts":"2021-06-14T10:57:32.138Z","caller":"aws-k8s-agent/main.go:41","msg":"Testing communication with server"}
{"level":"info","ts":"2021-06-14T10:57:32.139Z","caller":"aws-k8s-agent/main.go:41","msg":"Successful communication with the Cluster! Cluster Version is: v1.19+. git version: v1.19.8-eks-96780e. git tree state: clean. commit: 96780e1b30acbf0a52c38b6030d78531233456677. platform: linux/amd64"}
{"level":"debug","ts":"2021-06-14T10:57:32.141Z","caller":"ipamd/ipamd.go:292","msg":"Discovered region: eu-central-1"}
.
.
.
{"level":"debug","ts":"2021-06-14T10:57:32.341Z","caller":"awsutils/awsutils.go:584","msg":"Found ENI: eni-02bf67281234556, MAC 06:g8:6a:6d:f2:69, device 0"}
{"level":"debug","ts":"2021-06-14T10:57:32.342Z","caller":"awsutils/awsutils.go:604","msg":"Found CIDR 10.255.0.0/22 for ENI 06:g8:6a:6d:f2:69"}
{"level":"debug","ts":"2021-06-14T10:57:32.342Z","caller":"awsutils/awsutils.go:604","msg":"Found IP addresses [10.255.1.129] on ENI 06:g8:6a:6d:f2:69"}
{"level":"debug","ts":"2021-06-14T10:58:30.147Z","caller":"wait/wait.go:133","msg":"Checking for leaked AWS CNI ENIs."}
    ==========IPAM Initialization starts again===========
{"level":"info","ts":"2021-06-14T10:59:11.771Z","caller":"aws-k8s-agent/main.go:28","msg":"Starting L-IPAMD v1.7.10  ..."}
{"level":"info","ts":"2021-06-14T10:59:11.823Z","caller":"aws-k8s-agent/main.go:41","msg":"Testing communication with server"}
{"level":"info","ts":"2021-06-14T10:59:11.824Z","caller":"aws-k8s-agent/main.go:41","msg":"Successful communication with the Cluster! Cluster Version is: v1.19+. git version: v1.19.8-eks-96780e. git tree state: clean. commit: 96780e1b30acbf0a52c38b6030d78531233456677. covered region: eu-central-1platform: linux/amd64"}
```
/var/log/aws-routed-eni/plugin.log meanwhile has the following entries:
```
{"level":"info","ts":"2021-06-14T10:59:16.513Z","caller":"routed-eni-cni-plugin/cni.go:249","msg":"Received CNI del request: ContainerID(8b7dceb953454aa6888661823ec9014dad3588558d920cc74dac2598f438a992) Netns() IfName(eth0) Args(IgnoreUnknown=1;K8S_POD_NAMESPACE=default;K8S_POD_NAME=k8s-netact-871d12c4-5959-4e54-84b1-9ee358f9f275-kbz8q-196h9;K8S_POD_INFRA_CONTAINER_ID=8b7dceb953454aa6888661823ec9014dad3588558d920cc74dac2598f438a992) Path(/opt/cni/bin) argsStdinData({\"cniVersion\":\"0.3.1\",\"mtu\":\"9001\",\"name\":\"aws-cni\",\"pluginLogFile\":\"/var/log/aws-routed-eni/plugin.log\",\"pluginLogLevel\":\"DEBUG\",\"type\":\"aws-cni\",\"vethPrefix\":\"eni\"})"}
{"level":"error","ts":"2021-06-14T10:59:16.514Z","caller":"routed-eni-cni-plugin/cni.go:249","msg":"Error received from DelNetwork gRPC call for container 8b7dceb953454aa6888661823ec9014dad3588558d920cc74dac2598f438a992: rpc error: code = Unavailable desc = connection error: desc = \"transport: Error while dialing dial tcp 127.0.0.1:50051: connect: connection refused\""}
```
As per the entries in /var/log/aws-routed-eni/ipamd.log the IPAMD initialization process always gets to this point below and restarts.
```
{"level":"debug","ts":"2021-06-14T10:58:30.147Z","caller":"wait/wait.go:133","msg":"Checking for leaked AWS CNI ENIs."}
```
Looking into VPC CNI code the above message stems from the function [cleanUpLeakedENIs()](https://github.com/aws/amazon-vpc-cni-k8s/blob/dba11f45b747350cda623b08c6510876e2fb3f59/pkg/awsutils/awsutils.go#L1449-L1470)

>log.Debug("Checking for leaked AWS CNI ENIs.")
    networkInterfaces, err := cache.getFilteredListOfNetworkInterfaces()

Referring to the next function call for [getFilteredListOfNetworkInterfaces()](https://github.com/aws/amazon-vpc-cni-k8s/blob/dba11f45b747350cda623b08c6510876e2fb3f59/pkg/awsutils/awsutils.go#L1506-L1571)
getFilteredListOfNetworkInterfaces()- This function is calling DescribeNetworkInterfaces to get all available ENIs that were allocated by the AWS CNI plugin, but were not deleted.
>input := &ec2.DescribeNetworkInterfacesInput{
        Filters:    []*ec2.Filter{tagFilter, statusFilter},
        MaxResults: aws.Int64(describeENIPageSize),
    }

Can we possibly introduce a logic within the code to also check for a response from the EC2 service endpoint. The logic could work similar to what it is right now, but parsing for the response code *Exiting with rc 255* from failed call as seen in below output to push an error message about service endpoint being not available.

> aws ec2 describe-network-interfaces --region us-east-1 --debug
```
EndpointConnectionError: Could not connect to the endpoint URL: "https://ec2.us-east-1.amazonaws.com/"
2021-07-23 18:48:05,837 - MainThread - awscli.clidriver - DEBUG - Exiting with rc 255
 
Could not connect to the endpoint URL: "https://ec2.us-east-1.amazonaws.com/"
```

**Comment by jayanthvn** [comment]
@skkumsi 

Thanks for the detailed analysis. I will look into this but also happy to take PRs :)

**Comment by jayanthvn** [comment]
We do have a warning logged if the function errors out - https://github.com/aws/amazon-vpc-cni-k8s/blob/dba11f45b747350cda623b08c6510876e2fb3f59/pkg/awsutils/awsutils.go#L1458. Will have to repro this locally.

**Comment by Shreya027** [comment]
Hi, 
I tried reproducing the issue on v1.7.10 and v1.8.0, however, I was able to get the error when the EC2 endpoint wasn't reachable on DescribeNetworkInterfaces call. Check logs for v1.8.0 below:  
```
{"level":"debug","ts":"2021-11-18T20:10:05.171Z","caller":"awsutils/awsutils.go:516","msg":"Found ENI: eni-xxxx MAC xxxx, device x"} 
{"level":"error","ts":"2021-11-18T20:11:26.425Z","caller":"ipamd/ipamd.go:363","msg":"Failed to call ec2:DescribeNetworkInterfaces for [eni-xxx]: Re questError: send request failed\ncaused by: Post \"https://ec2.us-west-2.amazonaws.com/\": dial tcp xxxx connect: no route to host"} 
{"level":"info","ts":"2021-11-18T20:11:44.194Z","caller":"aws-k8s-agent/main.go:28","msg":"Starting L-IPAMD v1.8.0 ..."}
```

I checked to see if there is any EC2  timeout issue that could be possibly prevent the error from getting logged in the case you presented. I checked this [PR](https://github.com/aws/amazon-vpc-cni-k8s/pull/1370/files) that was merged in release v1.7.9, the PR has a timeout configured for the aws ec2 session [here](https://github.com/aws/amazon-vpc-cni-k8s/blob/a5169d8dab9e8c4918afdf098b9c1131ef9b0400/pkg/awsutils/awssession/session.go#L44). At startup, IPAMD uses ec2SVC call [here](https://github.com/aws/amazon-vpc-cni-k8s/blob/dba11f45b747350cda623b08c6510876e2fb3f59/pkg/awsutils/awsutils.go#L1623) for DescribeNetworkIntefacesCall. The session for ec2SVC initialised [here](https://github.com/aws/amazon-vpc-cni-k8s/blob/dba11f45b747350cda623b08c6510876e2fb3f59/pkg/awsutils/awsutils.go#L332) uses the timeout configured in PR. So I'm thinking the EC2 timeouts are properly configured and shouldn't cause the issue.

However, I figured out another timeout scenario apart from EC2 timeout that I need to explore further. The following call function ```go wait.Forever(cache.cleanUpLeakedENIs, time.Hour)```  [here](https://github.com/aws/amazon-vpc-cni-k8s/blob/dba11f45b747350cda623b08c6510876e2fb3f59/pkg/awsutils/awsutils.go#L340) triggers calling ```cache.cleanUpLeakedENIs``` periodically every hour for ever. ```cache.cleanUpLeakedENIs```  has functions that in turn call DescribeNetworkInterfaces.  Although ```wait.Forever```  is triggered for ever every hour, the timeout configured in the EC2 DescribeNetworkInterfaces call should atleast print this ```		log.Warnf("Unable to get leaked ENIs: %v", err)```  thrown [here](https://github.com/aws/amazon-vpc-cni-k8s/blob/aad0415a10dc4d1a6cda2bbc6c8cbbe856068e03/pkg/awsutils/awsutils.go#L1603)
